File: 织时_Android原型/app/src/main/java/com/vsa/visualsemanticagent/utils/NoticeSegmentExtractor.kt
package com.vsa.visualsemanticagent.utils

object NoticeSegmentExtractor {

    private val strongDateMarker = Regex(
        """(?:周[一二三四五六日天]|星期[一二三四五六日天]|今天|明天|后天|本周[一二三四五六日天]?|下周[一二三四五六日天]?|\d{1,2}月\d{1,2}日|\d{4}[./-]\d{1,2}[./-]\d{1,2})"""
    )

    private val timeMarker = Regex(
        """(?:\d{1,2}:\d{2}|\d{1,2}点(?:\d{1,2}分?)?|\d{1,2}时(?:\d{1,2}分?)?)"""
    )

    fun extractIndependentScheduleSegments(sourceText: String): List<String> {
        val normalized = sourceText
            .replace("\r\n", "\n")
            .replace('\u3000', ' ')
            .trim()
        if (normalized.isBlank()) return emptyList()

        val segments = splitByLines(normalized)
            .flatMap(::splitByInlineDates)
            .map(::normalizeSegment)
            .filter { it.isNotBlank() && containsTimeContext(it) }
            .distinct()

        return segments.takeIf { it.size >= 2 } ?: emptyList()
    }

    private fun splitByLines(sourceText: String): List<String> {
        val rawLines = sourceText
            .split('\n')
            .map { it.trim() }
            .filter { it.isNotBlank() }
        if (rawLines.size <= 1) return listOf(sourceText)

        val grouped = mutableListOf<String>()
        val current = StringBuilder()
        rawLines.forEach { line ->
            val startsNewBlock = containsStrongDate(line) && current.isNotBlank() && containsStrongDate(current.toString())
            if (startsNewBlock) {
                grouped += current.toString().trim()
                current.clear()
            }
            if (current.isNotBlank()) {
                current.append('\n')
            }
            current.append(line)
        }
        if (current.isNotBlank()) {
            grouped += current.toString().trim()
        }
        return grouped.ifEmpty { listOf(sourceText) }
    }

    private fun splitByInlineDates(segment: String): List<String> {
        val matches = strongDateMarker.findAll(segment).toList()
        if (matches.size < 2) return listOf(segment)

        val sharedPrefix = segment.substring(0, matches.first().range.first)
            .trim()
            .trimEnd('：', ':', '-', ' ')

        return matches.mapIndexed { index, match ->
            val start = match.range.first
            val endExclusive = matches.getOrNull(index + 1)?.range?.first ?: segment.length
            val slice = segment.substring(start, endExclusive)
                .trim()
                .trimStart('，', ',', '；', ';', '。', '、')
                .trimEnd('，', ',', '；', ';', '。', '、')
            normalizeSegment(
                when {
                    sharedPrefix.isBlank() -> slice
                    slice.startsWith(sharedPrefix) -> slice
                    else -> "$sharedPrefix $slice"
                }
            )
        }
    }

    private fun normalizeSegment(segment: String): String {
        return segment
            .replace("\n", " ")
            .replace(Regex("""\s+"""), " ")
            .trim()
    }

    private fun containsStrongDate(text: String): Boolean = strongDateMarker.containsMatchIn(text)

    private fun containsTimeContext(text: String): Boolean {
        return containsStrongDate(text) || timeMarker.containsMatchIn(text)
    }
}

File: timeweaver_flutter/lib/utils/notice_segment_extractor.dart
class NoticeSegmentExtractor {
  static final RegExp _strongDateMarker = RegExp(
    r'(?:周[一二三四五六日天]|星期[一二三四五六日天]|今天|明天|后天|本周[一二三四五六日天]?|下周[一二三四五六日天]?|\d{1,2}月\d{1,2}日|\d{4}[./-]\d{1,2}[./-]\d{1,2})',
  );
  static final RegExp _timeMarker = RegExp(
    r'(?:\d{1,2}:\d{2}|\d{1,2}点(?:\d{1,2}分?)?|\d{1,2}时(?:\d{1,2}分?)?)',
  );

  static List<String> extractIndependentScheduleSegments(String sourceText) {
    final normalized = sourceText
        .replaceAll('\r\n', '\n')
        .replaceAll('\u3000', ' ')
        .trim();
    if (normalized.isEmpty) return const [];

    final segments = _splitByLines(normalized)
        .expand(_splitByInlineDates)
        .map(_normalizeSegment)
        .where((segment) => segment.isNotEmpty && _containsTimeContext(segment))
        .toSet()
        .toList();
    return segments.length >= 2 ? segments : const [];
  }

  static List<String> _splitByLines(String sourceText) {
    final lines = sourceText
        .split('\n')
        .map((line) => line.trim())
        .where((line) => line.isNotEmpty)
        .toList();
    if (lines.length <= 1) return [sourceText];

    final grouped = <String>[];
    final current = StringBuffer();
    for (final line in lines) {
      final currentText = current.toString();
      final startsNewBlock =
          _containsStrongDate(line) &&
          currentText.trim().isNotEmpty &&
          _containsStrongDate(currentText);
      if (startsNewBlock) {
        grouped.add(currentText.trim());
        current.clear();
      }
      if (current.toString().trim().isNotEmpty) current.write('\n');
      current.write(line);
    }
    if (current.toString().trim().isNotEmpty) {
      grouped.add(current.toString().trim());
    }
    return grouped.isEmpty ? [sourceText] : grouped;
  }

  static List<String> _splitByInlineDates(String segment) {
    final matches = _strongDateMarker.allMatches(segment).toList();
    if (matches.length < 2) return [segment];

    final sharedPrefix = segment
        .substring(0, matches.first.start)
        .trim()
        .replaceFirst(RegExp(r'[：:\-\s]+$'), '');
    return List.generate(matches.length, (index) {
      final start = matches[index].start;
      final end = index + 1 < matches.length
          ? matches[index + 1].start
          : segment.length;
      final slice = segment
          .substring(start, end)
          .trim()
          .replaceFirst(RegExp(r'^[，,；;。、]+'), '')
          .replaceFirst(RegExp(r'[，,；;。、]+$'), '');
      return _normalizeSegment(
        sharedPrefix.isEmpty || slice.startsWith(sharedPrefix)
            ? slice
            : '$sharedPrefix $slice',
      );
    });
  }

  static String _normalizeSegment(String segment) {
    return segment.replaceAll('\n', ' ').replaceAll(RegExp(r'\s+'), ' ').trim();
  }

  static bool _containsStrongDate(String text) {
    return _strongDateMarker.hasMatch(text);
  }

  static bool _containsTimeContext(String text) {
    return _containsStrongDate(text) || _timeMarker.hasMatch(text);
  }
}
